src/compiler/checker.ts

```diff
@@ -7723,27 +7730,27 @@ namespace ts {
             if (types.length === 0) {
                 return emptyObjectType;
             }
-            const typeSet = [] as TypeSet;
-            addTypesToIntersection(typeSet, types);
-            if (typeSet.containsNever) {
+            const typeSet: Type[] = [];
+            const includes = addTypesToIntersection(typeSet, 0, types);
+            if (includes & TypeIncludes.Never) {
                 return neverType;
             }
-            if (typeSet.containsAny) {
+            if (includes & TypeIncludes.Any) {
                 return anyType;
             }
-            if (typeSet.containsEmptyObject && !typeSet.containsObjectType) {
+            if (includes & TypeIncludes.EmptyObject && !(includes & TypeIncludes.ObjectType)) {
                 typeSet.push(emptyObjectType);
             }
             if (typeSet.length === 1) {
                 return typeSet[0];
             }
-            const unionIndex = typeSet.unionIndex;
-            if (unionIndex !== undefined) {
+            if (includes & TypeIncludes.Union) {
                 // We are attempting to construct a type of the form X & (A | B) & Y. Transform this into a type of
                 // the form X & A & Y | X & B & Y and recursively reduce until no union type constituents remain.
+                const unionIndex = findIndex(typeSet, t => (t.flags & TypeFlags.Union) !== 0);

```

Comment:
Rather than calculating this all the time (which could get expensive for larger intersections), could we not reserve multiple bits of the flags for the `UnionIndex` (all the remaining free bits, even), set them with the index as types are added like before, and only perform this lookup if it's saturated (indicating that all we know when it's saturated is that the index is > or = the size of the saturated number, but then at least we can start searching from there)? Would prevent us from needing to do this lookup on intersections where a union appears within the first 2^20-1 or so members, and let us skip those first members if it's after them. This probably won't be noticable until you started defining a generic intersection with a few hundred plus elements where one of the last few elements is occasionally a union, but still.

Reply:
Yeah, I considered that, but compared to the amount of work we do to actually distribute over the union type it, and the relative rarity of this operation in the first place, it just seemed like extra complexity.